stable-2.10.0 (#5859)
linkerd/linkerd2: CHANGES.md

# Changes

## stable-2.10.0

This release introduces Linkerd extensions. The default control plane no longer
includes Prometheus, Grafana, the dashboard, or several other components that
previously shipped by default.  This results in a much smaller and simpler set

> of core functionality.  Visibility and metrics functionality is now available

Reviewer: Want to use plural here since it's a set of functionalities.

```suggestion
of core functionalities.  Visibility and metrics functionality is now available
```

I think it should also be "Visibility and metrics functionalities are now..", but that stands out less to me.

in the Viz extension under the `linkerd viz` command.  Cross-cluster

> communication functionality is now available in the Multicluster extensions

Reviewer: ```suggestion
communication functionality is now available in the Multicluster extension
```

now available in the Jaeger extension under the `linkerd jaeger` command.

This release also introduces the ability to mark certain ports as "opaque",
indicating that the proxy should treat the traffic as opaque TCP instead of
attempting protocol detection.  This allows the proxy to provide TCP metrics
and mTLS for server-speaks-first protocols.  It also enables support for
TCP traffic in the Multicluster extension.

**Upgrade notes**: Please see the [upgrade
instructions](https://linkerd.io/2/tasks/upgrade/#upgrade-notice-stable-2100).

* Proxy
  * Updated the proxy to use TLS version 1.3; support for TLS 1.2 remains
    enabled for compatibility with prior proxy versions
  * Improved support for server-speaks-first protocols by allowing ports to be
    marked as opaque, causing the proxy to skip protocol detection.  Ports can
    be marked as opaque by setting the `config.linkerd.io/opaque-ports`
    annotation on the Pod and Service or by using the `--opaque-ports` flag with
    `linkerd inject`
  * Ports `25,443,587,3306,5432,11211` have been removed from the default skip
    ports; all traffic through those ports is now proxied and handled opaquely

>     by default

Reviewer 2: Maybe worth linking to the opaque ports doc from here

>   * Fixed an issue that could cause the inbound proxy to fail meshed HTTP/1
>     requests from older proxies (from the stable-2.8.x vintage)

Reviewer 2: This issue was fixed in 2.9.4. However:

```suggestion
  * Fixed an issue that could cause proxies in "ingress mode"
    (`linkerd.io/inject: ingress`) to use an excessive amount of memory.
  * Improved diagnostic logging around "fail fast" and "max-concurrency
    exhausted" error messages 
```

  * Added a new `/shutdown` admin endpoint that may only be accessed over the
    loopback network allowing batch jobs to gracefully terminate the proxy on

>     completion

Reviewer 2: Maybe worth a callout to Linkerd-await, which calls this.

    config map or secret (thanks @mgoltzsche!)
  * Added PodDisruptionBudgets to the control plane components so that they
    cannot be all terminated at the same time during disruptions
    (thanks @tustvold!)

* CLI
  * Changed the `check` command to include each installed extension's `check`
    output; this allows users to check for proper configuration and installation
    of Linkerd without running a command for each extension
  * Moved the `metrics`, `endpoints`, and `install-sp` commands into subcommands
    under the `diagnostics` command
  * Added an `--opaque-ports` flag to `linkerd inject` to easily mark ports
    as opaque.
  * Added the `repair` command which will repopulate resources needed for
    properly upgrading a Linkerd installation
  * Added Helm-style `set`, `set-string`, `values`, `set-files` customization
    flags for the `linkerd install` and `linkerd upgrade` commands
  * Introduced the `linkerd identity` command, used to fetch the TLS certificates
    for injected pods (thanks @jimil749)
  * Removed the `get` and `logs` command from the CLI

* Helm
  * Changed many Helm values, please see the upgrade notes

* Viz

>   * Updated the Web UI to only display the "Gateway" sidebar link when the

Reviewer 2: Probably worth calling out that the CLI now has `linkerd viz` set of subcommands, including `linkerd viz install`.

    multicluster extension is active
  * Added a `linkerd viz list` command to list pods with tap enabled
  * Fixed an issue where the `tap` APIServer would not refresh its certs
    automatically when provided externally—like through cert-manager

* Multicluster
  * Added support for cross-cluster TCP traffic
  * Updated the service mirror controller to copy the
    `config.linkerd.io/opaque-ports` annotation when mirroring services so that
    cross-cluster traffic can be correctly handled as opaque
  * Added support for multicluster gateways of types other than LoadBalancer
    (thanks @DaspawnW!)

* Jaeger

>   * Added a `linkerd jaeger list` command to list pods with tracing enabled

Reviewer 2: Similarly, `linkerd jaeger install`

* Other

>   * Docker images are now hosted on the `cr.l5d.io` registry

Reviewer 2: Technically not new, since 2.9.4 is hosted on this registry.

# Changes

## stable-2.10.0

This release introduces Linkerd extensions. The default control plane no longer
includes Prometheus, Grafana, the dashboard, or several other components that

> previously shipped by default.  This results a much smaller and simpler set

Reviewer 3: Not totally sure about this one:
```suggestion
previously shipped by default.  This results in a much smaller and simpler set
```

of core functionality.  Visibility and metrics functionality is now available
in the Viz extension under the `linkerd viz` command.  Cross-cluster
communication functionality is now available in the Multicluster extensions
under the `linkerd multicluster` command.  Distributed tracing functionality is

> not available in the Jaeger extension under the `linkerd jaeger` command.

Reviewer 3: ```suggestion
now available in the Jaeger extension under the `linkerd jaeger` command.
```

    ports; all traffic through those ports is now proxied and handled opaquely
    by default
  * Fixed an issue that could cause the inbound proxy to fail meshed HTTP/1
    requests from older proxies (from the stable-2.8.x vintage)
  * Added a new `/shutdown` admin endpoint that may only be accessed over the
    loopback network allowing batch jobs to gracefully terminate the proxy on
    completion

* Control Plane
  * Removed all components and functionality related to visibility, tracing,
    or multicluster.  These have been moved into extensions
  * Changed the identity controller to receive the trust anchor via environment
    variable instead of by flag; this allows the certificate to be loaded from a
    config map or secret (thanks @mgoltzsche!)
  * Added PodDisruptionBudgets to the control plane components so that they
    cannot be all terminated at the same time during disruptions
    (thanks @tustvold!)
  * Added missing label to the `linkerd-config-overrides` secret to avoid
    breaking upgrades performed with the help of `kubectl apply --prune`
  * Fixed an issue where the `proxy-injector` and `sp-validator` did not refresh

>     their certs automatically when provided externally—like through cert-manager

Reviewer 3: These two points are already in 2.9.x, so we can omit them here.

* CLI
  * Changed the `check` command to include each installed extension's `check`
    output; this allows users to check for proper configuration and installation
    of Linkerd without running a command for each extension
  * Moved the `metrics`, `endpoints`, and `install-sp` commands into subcommands
    under the `diagnostics` command
  * Added an `--opaque-ports` flag to `linkerd inject` to easily mark ports
    as opaque.
  * Added the `repair` command which will repopulate resources needed for
    properly upgrading a Linkerd installation
  * Added Helm-style `set`, `set-string`, `values`, `set-files` customization

>     flags for the `linkerd install` command

Reviewer 3: ```suggestion
    flags for the `linkerd install` and `linkerd upgrade` commands
```
